kernel/main.c

```diff
@@ -340,6 +353,52 @@ async_init(void)
       panic("Unable to create a kthread for do_async_init()");
 }
 
+static void
+setup_uefi_runtime_services() {
+   EFI_RUNTIME_SERVICES *RT = TO_PTR(uefi_rt_addr);
+
+   printk("Runtime Services: %p\n", RT);
+
+   struct mem_region ma;
+
+   EFI_MEMORY_DESCRIPTOR *virt_map = kmalloc((size_t)get_mem_regions_count());
+   int num_entries_virt_map = 0;
+
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+      get_mem_region(i, &ma);
+
+      if (ma.type == TILCK_BOOT_EFI_RUNTIME_CODE_REQ || ma.type == TILCK_BOOT_EFI_RUNTIME_DATA_REQ) {
+         EFI_MEMORY_DESCRIPTOR desc;
+
+         desc.Type = ma.type == TILCK_BOOT_EFI_RUNTIME_CODE_REQ ? EfiRuntimeServicesCode : EfiRuntimeServicesData;
+         desc.PhysicalStart = (ulong)ma.addr;
+         desc.VirtualStart = (ulong)KERNEL_PA_TO_VA(ma.addr);
+         desc.NumberOfPages = ma.len / PAGE_SIZE;
+         desc.Attribute = EFI_MEMORY_RUNTIME;
+
+         virt_map[num_entries_virt_map] = desc;
+         num_entries_virt_map++;
+
+         printk("Type: %d\tPhysical Address: %lld\t Virtual Address: %lld\n", desc.Type, desc.PhysicalStart, desc.VirtualStart);
+      }
+   }
+
+   printk("Virtual map: %p\n", virt_map);
+
+   printk("Time %p\n", RT->GetTime);
+   /*
+   printk("SVAM %p\n", RT->SetVirtualAddressMap);
+
+   UINTN mem_map_size = num_entries_virt_map * sizeof(EFI_MEMORY_DESCRIPTOR);
+   UINTN desc_size = sizeof(EFI_MEMORY_DESCRIPTOR);
+   UINT32 desc_version = EFI_MEMORY_DESCRIPTOR_VERSION;
+
+   EFI_STATUS status = RT->SetVirtualAddressMap(mem_map_size, desc_size, desc_version, virt_map);
```

Comment:
Unfortunately, after the last conversation, it looks like that calling `SetVirtualAddressMap()` like this here won't work because the UEFI runtime mem regions are not identity-mapped. 

You might need to call `map_pages()` in the loop, for each of those regions, with VADDR = PADDR. See how `linear_map_mem_region()` uses that function. Then, here at the end, after the call succeeded, you'll need to re-iterate that loop and call: unmap_pages() for each region.